command/init.go

```diff
@@ -163,6 +177,31 @@ func (c *InitCommand) Run(args []string) int {
 	return 0
 }
 
+// load the complete module tree, and fetch any missing providers
+func (c *InitCommand) getProviders(path string) error {
+	mod, err := c.Module(path)
+	if err != nil {
+		return err
+	}
+
+	if err := mod.Validate(); err != nil {
+		return err
+	}
+
+	requirements := terraform.ModuleTreeDependencies(mod, nil).AllPluginRequirements()
```

Comment:
oh yes, thanks! I forgot that we should now load state here. 

hmm, that's a good point, but I think it's an edge case we can tackle later. I don't see it being very common that a named state will have resources from an unnamed provider, and some good UI feedback about why `init` is required may be all we need.